### ISSUE TITLE
[$2000] The scrollbar does not scroll properly when long pressing up or down arrow key in RHN

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#26403]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open search by pressing CTRL + K.
2. Long press up or down arrow key.
3. Observe the scrollbar in search option list.

## Expected Result:
The scrollbar should be scroll properly.

## Actual Result:
The scrollbar does not scroll properly.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** v1.3.60-1

**Reproducible in staging?:** Y

**Reproducible in production?:** Y

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/115492554/3ba35a1e-bd97-4cb4-b513-55699cf112b8

https://github.com/Expensify/App/assets/115492554/047a8655-c795-4298-9811-45f67bbf99bc

**Expensify/Expensify Issue URL:**

**Issue reported by:** @@oleksandr-pantsyr

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1692725411476829

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01701d970faa399749</li>
        <li>Upwork Job ID: 1699130136055246848</li>
        <li>Last Price Increase: 2023-10-27</li>
    </ul>
</details>

**Comment by kanno927** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The scrollbar does not scroll properly when long pressing up or down arrow key in RHN

### What is the root cause of that problem?
Please take a look at this code.
https://github.com/Expensify/App/blob/main/src/components/OptionsSelector/BaseOptionsSelector.js#L259-L261
https://github.com/Expensify/App/blob/main/src/components/OptionsSelector/BaseOptionsSelector.js#L386-L392

The updateFocusedIndex method is called when the user clicks the up or down arrow buttons.
When users press and hold the up and down arrow buttons, the updateFousedIndex function is called several times, but only the last called function works. So the list doesn't scroll one by one.
The above issue causes the list to scroll all at once to the last index.

### What changes do you think we should make in order to solve the problem?
So, while the list is scrolling to one item, we need to prevent the key press event.

### What alternative solutions did you explore? (Optional)
``` js
constructor(props) {
  //...other code...
        
  //add this code
  this.isScrolling = false;
}

//update this method
updateFocusedIndex(index) {
  if(this.isScrolling) {
    return;
  }
  this.setState({focusedIndex: index}, () => this.scrollToIndex(index));
}

scrollToIndex(index, animated = true) {
  //...other code...

  //add this code
  this.isScrolling = true
  setTimeout(() => {
    this.isScrolling = false
  }, 10); 
}
```

Result:
https://github.com/Expensify/App/assets/122101279/252247e3-00aa-45c3-8e03-8919b1ceb575

**Comment by slafortune** [comment]
report looks good, applying the external label

**Comment by slafortune** [comment]
@Ollyws what do you think of the proposal presented by @kanno927 ?

Adding a new BZ member, I am out on sabbatical.

**Comment by kanno927** [comment]
@Ollyws Thanks for your kind message.

## Proposal
[Updated](https://github.com/Expensify/App/issues/26403#issuecomment-1702357271)

**Comment by kevinksullivan** [comment]
Hi @Ollyws what's the latest on the review?

**Comment by kevinksullivan** [comment]
Waiting on proposals for now.

**Comment by kevinksullivan** [comment]
doubled the price

**Comment by AmjedNazzal** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The scrollbar does not scroll properly when long pressing up or down arrow key in RHN

### What is the root cause of that problem?

The issue here is that the arrow up and down events and the updating of focusedIndex state variable are not in sync with the animation we are applying to the scroll, essentially the scrolling animation is getting called over and over again and not giving it a chance to catch up with the currentIndex it's supposed to scroll to, animations in general behave like a delay so something delayed will not work well in conjunction with something instant, leading to buggy behaviour.

### What changes do you think we should make in order to solve the problem?

We have two options here without having to refractor or do very major changes, the first option being that we could just let go of the scrolling animation:

```javascript
scrollToIndex(index, animated = false) {
```
The following video would be how the scrolling will look without the animation: 

https://github.com/Expensify/App/assets/74202751/21d11285-0ecb-4e1e-b749-64740dfec265


The second option is a bit more convoluted, what we can do is use a state variable to block additional calls to scrollToIndex and updating focusedIndex for a duration of time to give the animation enough time to finish, additionally, the index changes to 0 when we scroll down from the last item and changes to the last item when we scroll up from item 0 which causes a very long animation of scrolling from one end to another, this is problematic in this case so we need to disable animation when we are doing this jump from one end to another:

```javascript
updateFocusedIndex(index) {
    const isJumping = (this.state.focusedIndex === 0 && index === this.state.allOptions.length - 1) || (this.state.focusedIndex === this.state.allOptions.length - 1 && index === 0);
    if (this.state.isScrolling) {
        return;
    }
    this.setState({focusedIndex: index}, () => this.scrollToIndex(index, !isJumping));
}
```

```javascript
scrollToIndex(index, animated = true) {
    ...
    this.setState({ isScrolling: true }, () => {
        this.list.scrollToLocation({
            sectionIndex: adjustedSectionIndex, 
            itemIndex, 
            animated
        });

        setTimeout(() => {
            this.setState({ isScrolling: false });
          }, 200); 
        });
```

Understandably this is not an ideal fix, however if we want to not use any kind of timeout or debounce we would need to do quite a bit of refractoring as SectionList doesn't offer much control for animation or any sort of an indication for when scrolling is done

### Result

https://github.com/Expensify/App/assets/74202751/8585db24-2523-46ea-b4ab-69db45acaddc

**Comment by AmjedNazzal** [comment]
@Ollyws You are right, Firefox behaves differently and so does Safari as shown in the video below: 

https://github.com/Expensify/App/assets/74202751/d5ce4392-86f1-4737-b68c-44c3a8ce7a32



The idea here is that each browser's engine will handle the animation queue differently leading to inconsistency accross browsers. 

It seems like there are three options, the first being turning off animation completely, the second is that we can create a scroll helper function, we can use onScroll event returned from the list with a custom scrolling handler to keep track of the offset and determine whether or not to allow new scrolling calls, for example: 

```javascript
handleOnScroll(event) {
        this.setState({ isScrolling: true, scrollLimit: false })
        
        const newOffset = event.nativeEvent.contentOffset.y;
        const scrollLimit = event.nativeEvent.contentSize.height - event.nativeEvent.layoutMeasurement.height;
        
        if (Math.round(newOffset) === this.state.targetOffset) { 
            this.setState({ isScrolling: false });
        }
        if( this.state.targetOffset >= scrollLimit || this.state.focusedIndex <= 2 ){
            this.setState({ scrollLimit: true });
        }
        if(!this.state.isScrolling && !this.state.scrollLimit){
            this.scrollToIndex(this.state.focusedIndex)
        }  
      }
```

```javascript
scrollToIndex(index, animated = true) {
...
const offset = _.get(this.targetItem, [itemIndex, 'offset'], 0);
        if (this.state.isScrolling && !this.state.scrollLimit) {
            return;
        }
        this.setState({ targetOffset: offset }, () => {
            this.list.scrollToLocation({sectionIndex: adjustedSectionIndex, itemIndex, animatedState});
        });
```

The third option would be to delay scrolling as an animation mimic as I've mentioned previously.

**Comment by AmjedNazzal** [comment]
@Ollyws No what we are blocking in the second option is the scrolling, the button press is unaffected and the focusedIndex would change normally but the scrolling would try to catch up whenever the animation is done.

**Comment by kevinksullivan** [comment]
@Ollyws what's the latest on moving this forward?

**Comment by AmjedNazzal** [comment]
@Ollyws I apologize I've been really wanting to look further into this and see how we could refine and improve a helper function that would work well but unfortunately as you can see in the video below, since pulling main recently the app and scrolling have been so laggy for me that it's been impossible to do any kind of testing. I don't know why this is happening but it made looking into this impossible for me.


https://github.com/Expensify/App/assets/74202751/da42017c-6d5d-4582-a4ac-593791d975a6

**Comment by kevinksullivan** [comment]
Doubling

**Comment by s-alves10** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Scroll doesn't work properly when long pressing arrow up or down key in chat search page

### What is the root cause of that problem?

When long-pressing arrow up or down key, arrow up/down event handlers are called in a quick succession
https://github.com/Expensify/App/blob/4e722074e8d6efc85cbf0bf456f09e89e23d6e59/src/components/OptionsSelector/BaseOptionsSelector.js#L292

This function changes the focused index and scroll to the index. Scrolling animation is slow, and the function takes tens of milliseconds in general. But the keyboard event occurs more quickly. So before the `updateFocusedIndex` call is finished, other `updateFocusedIndex` calls are made.

State update is relatively faster than animation
This is the root cause

### What changes do you think we should make in order to solve the problem?

We need to prevent the above function from being called while it is being executed.
Debounce the arrow key handlers in [ArrowKeyFocusManager](https://github.com/Expensify/App/blob/main/src/components/ArrowKeyFocusManager.js)

Update 
https://github.com/Expensify/App/blob/533a373d9df451c389b2effbe4e6b487ecea43be/src/components/ArrowKeyFocusManager.js#L40-L41
to
```
// import {debounce} from 'lodash';

        this.onArrowUpKey = debounce(this.onArrowUpKey, 10).bind(this);
        this.onArrowDownKey = debounce(this.onArrowDownKey, 10).bind(this);
```

Debouncing the arrow key handlers would be good from the view of UX


<details>
    <summary>Result</summary>


https://github.com/Expensify/App/assets/126839717/51910c9e-788f-46e0-9bef-b5beb85430fd


</details>

### What alternative solutions did you explore? (Optional)

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/26403#issuecomment-1783511478)

@Ollyws 
I updated the proposal. The recorded video looks not smooth, but it isn't. Please test the solution in your machine. 
Thank you

**Comment by shubham1206agra** [comment]
@Ollyws I think this is performance optimization
Please check this sandbox
https://codesandbox.io/s/issue-react-native-for-web-forked-j46fpm?file=/src/App.js

**Comment by s-alves10** [comment]
I've tried reanimated scrollTo already, but didn't find considerable improvement

**Comment by shubham1206agra** [comment]
@Ollyws I think we should discuss what to do next about this in Slack. Maybe an engineer might help with this.

**Comment by s-alves10** [comment]
@Ollyws @shubham1206agra 

Will you test my updated solution and give me feedback, please?

**Comment by MahmudjonToraqulov** [comment]
# Proposal
## Please re-state the problem that we are trying to solve in this issue.
Scroll doesn't work properly when long pressing arrow up or down key in chat search page

## What is the root cause of that problem?
I think `animated` is causing the problem. Changing it into 'false' as default value may can help.
https://github.com/Expensify/App/blob/4e722074e8d6efc85cbf0bf456f09e89e23d6e59/src/components/OptionsSelector/BaseOptionsSelector.js#L301

## What changes do you think we should make in order to solve the problem?
We should make the change like: Turning `animated` from `true` to `false`

## Result
[screen-capture.webm](https://github.com/Expensify/App/assets/91727683/a588c011-88a4-41c7-911f-f18195ba66dd)

## What alternative solutions did you explore? (Optional)
N/A

**Comment by dragnoir** [comment]
I just spotted the problem. If you can wait a few more days for another proposal, I'd really appreciate it. I believe I've identified the issue.

**Comment by sourcecodedeveloper** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When search using RHN, the scrollbar does not scroll properly when long pressed up or down arrow keys.

### What is the root cause of that problem?
There are several areas of improvements. The search page is using section list which is setting renderItem without using useCallback or implementing shouldComponentUpdate. The list is being recreated for each render and for each item in the list. The scrollToLocation is dependent on adjustedSectionIndex and itemIndex which is hard to maintain. Also flattenSections() is flattening sections into an array, the sectionIndex is unnecessary to maintain.


### What changes do you think we should make in order to solve the problem?
Refactor renderItem from BaseOptionsList.js into another class component BaseOptionsView.js and implement shouldComponentUpdate. Currently, the renderItem is called for each item on scrolling whether the item is highlighted or not. The shouldComponentUpdate skips re-renders and only updates the necessary items.


https://github.com/Expensify/App/blob/a273354130bbbc057dee4eefb1f9ac0371bec26a/src/components/OptionsList/BaseOptionsList.js#L279

          shouldComponentUpdate(nextProps, nextState) {
            if (nextProps.focusedIndex === nextProps.index + nextProps.section.indexOffset) {
              return true;
            }
            if (this.state.previousFocusedIndex === nextProps.index + nextProps.section.indexOffset) {
              return true;
            }

            return false;
          }

Instead of using sectionIndex and ItemIndex use scrollIntoView. ScrollIntoView allows smooth scrolling and is not dependent on sectionIndex or itemIndex.


https://github.com/Expensify/App/blob/a273354130bbbc057dee4eefb1f9ac0371bec26a/src/components/OptionsSelector/BaseOptionsSelector.js#L341
    
        window.requestAnimationFrame(() => {

            const scroll = this.state.listRef.current.getScrollResponder();
            if (this.state.focusedIndex == this.state.lastItemIndex) {
                scroll.scrollToEnd({animated: true});
            }
            else if (this.state.focusedIndex == 0) {
                scroll.scrollTo({
                    y: 0,
                    animated: true,
                });
            }
            else {
                const element = this.state.listRef.current.props.extraData.current;
                if (!element) {
                    return;
                }
                element.scrollIntoView({behavior: 'smooth', block: 'center'});
            }
        });

The improvement shown is significantly better than that on staging.

[bug_solution.webm](https://github.com/Expensify/App/assets/141340971/2f883ff2-1548-4647-a54a-78a2d1629e68)

Staging Search

[staging_rhn.webm](https://github.com/Expensify/App/assets/141340971/9f83d280-88ef-4d43-8a21-d0f7052975d9)

**Comment by kevinksullivan** [comment]
asking here

https://expensify.slack.com/archives/C03UK30EA1Z/p1701124526317179

**Comment by sourcecodedeveloper** [comment]
@Ollyws the [proposal](https://github.com/Expensify/App/issues/26403#issuecomment-1823414985) uses smooth animation. I can prepare branch for testing if needed.

**Comment by kevinksullivan** [comment]
@Ollyws thoughts on the above? No word from expert agencies yet.

**Comment by sourcecodedeveloper** [comment]
> Will take a look asap.

@Ollyws please provide some feedback on [proposal](https://github.com/Expensify/App/issues/26403#issuecomment-1823414985) and share any ideas that you may have.

**Comment by sourcecodedeveloper** [comment]
@Ollyws Please get latest from [branch](https://github.com/sourcecodedeveloper/App/tree/optionsListRefactor)

**Comment by sourcecodedeveloper** [comment]
@Ollyws the shouldcomponentupdate interface would greatly improve the scrolling along with scrollintoview using smooth animation. The optionslist is rendering each item even if the item is not required to be rendered. The .debounce would add delay and not really improve the list scrolling. Please share feedback from testing proposal.

**Comment by sourcecodedeveloper** [comment]
@Ollyws the OptionsSelector does show message in console the list is large and slow to update and suggests to follow React performance best practices i.e. implement shouldComponentUpdate. window.requestAnimationFrame improved the smooth animation considerably on chrome when I tested long press scrolling. I can also include _.debounce in the proposal.

**Comment by dragnoir** [comment]
Hi @Ollyws _.debounce will not solve the issues.

Also window.requestAnimationFrame alone will not solve the bug.
As I mentioned, monitoring scroll completion via SectionList callbacks is Important too.

I tested  _.debounce and window.requestAnimationFrame without waiting for the section to scroll and the issue is always there.
I can share the full update if you want

Here's what I [changed ](https://github.com/dragnoir/App/commit/293345fb27105757aa1b120c7949646b372ca6b4)(not the perfect one and a not clean) 
but the waiting for the logic waiting for list to scroll is there.

**Comment by sourcecodedeveloper** [comment]
@Ollyws the other issue, the current scroll is using scrollToLocation, which uses scrollToIndex of VirtualizedList. The scrollIntoView() runs much smoother than scrollToIndex. The smooth animation would definitely benefit from the use of scrollIntoView() when using long press. Definitely consider scrollIntoView + shouldComponentUpdate to boost the scrolling and smooth animation. There are also other alternatives to requestAnimationFrame e.g. setImmediate.

**Comment by sourcecodedeveloper** [comment]
@Ollyws wondering if you are curious about the alternative solution. just going with _.debounce will certainly not solve the issue.

**Comment by sourcecodedeveloper** [comment]
@Ollyws Please review revised [proposal](https://github.com/Expensify/App/issues/26403#issuecomment-1823414985). I figured out clean and straightforward approach. The approach effectively resolves the issue and is simple to understand.

**Comment by sourcecodedeveloper** [comment]
I do not get the jittery effect, I tested on MacOS / Chrome. 
Video attached

[poc.webm](https://github.com/Expensify/App/assets/141340971/607f76da-e1c3-4e70-b2c1-a4db16fa85a7)

I am curious how come you are getting the jittery effect. could you kindly run the app from the branch below and post video.
https://github.com/sourcecodedeveloper/App/tree/scrollbarBehavior

**Comment by sourcecodedeveloper** [comment]
@Ollyws  I do not get the jittery effect, I tested on MacOS / Chrome. 
Video attached

[poc.webm](https://github.com/Expensify/App/assets/141340971/607f76da-e1c3-4e70-b2c1-a4db16fa85a7)

I am curious how come you are getting the jittery effect. could you kindly run the app from the branch below and post video.
https://github.com/sourcecodedeveloper/App/tree/scrollbarBehavior

**Comment by s-alves10** [comment]
I think combining `debounce` and `window.requestAnimationFrame` together would be a solution. They would help us control the scroll behavior and key-press behavior. As you can see below, this solution works perfect for HT account as well


https://github.com/Expensify/App/assets/126839717/167c686a-1a4b-4ef2-8bde-2e68a65b8af2

**Comment by sourcecodedeveloper** [comment]
@Ollyws I have attached video using high traffic account and I still do get the jittery effect. The search list that I used had only 1 section where as the search list used by the high traffic account had 2 sections. The scrollToLocation does depend on the sectionIndex, therefore if maybe the sectionIndexes are not being prepared properly because there were no issues at using search list using 1 section.

[poc.webm](https://github.com/Expensify/App/assets/141340971/e945ce49-8cb5-445b-b08c-b2880c4ec34e)

Can you re-test the proposal but use search list with only 1 section and then confirm the jittery effect. If you can confirm the behavior for search list with only 1 section that would greatly assist in narrowing the issue.

**Comment by kevinksullivan** [comment]
Hi @Ollyws can you provide an update here?

**Comment by shubham1206agra** [comment]
@kevinksullivan Can you put this issue on hold for https://github.com/Expensify/App/pull/31606?

**Comment by sourcecodedeveloper** [comment]
@Ollyws Please test branch and review revised proposal. I have used useRef and removed setState on focusedIndex from updateFocusedIndex. Using setState on long press causes re-render each time where as there is no re-render with useRef approach.

branch for testing proposal
https://github.com/sourcecodedeveloper/App/tree/scrollbarBehavior

Attached poc, I tested on chrome and do not get any jittery effect

[poc.webm](https://github.com/Expensify/App/assets/141340971/0deac4ca-7ac1-4c44-8bb3-e73dfe78ce16)

**Comment by sourcecodedeveloper** [comment]
@shubham1206agra I suggest to finish merging the PR https://github.com/Expensify/App/pull/31606 first as the issue is about migrating BaseOptionsSelector.js to function component, the scrollbar issue is separate from the migration issue. If after migration the scrollbar issue is also resolved then the scrollbar issue can be closed instead of put on hold.

**Comment by sourcecodedeveloper** [comment]
@Ollyws @kevinksullivan Awaiting response for revised [proposal](https://github.com/Expensify/App/issues/26403#issuecomment-1823414985).

**Comment by sourcecodedeveloper** [comment]
@Ollyws Please help test the useRef approach mentioned in the proposal.

**Comment by kevinksullivan** [comment]
sounds good, included that PR in the title.

**Comment by kevinksullivan** [comment]
still held

**Comment by dragnoir** [comment]
@kevinksullivan pls check https://github.com/Expensify/App/issues/16184#issuecomment-1971738161 and https://expensify.slack.com/archives/C01GTK53T8Q/p1709232289899589?thread_ts=1709156803.359359&cid=C01GTK53T8Q

The component we are holding for will not be used. Also the main issue here (search) already use SectionList and not OptionSelector.

Note: even after migrating to SectionList, we still see the issue. The scrollbar does not scroll properly.

**Comment by mallenexpensify** [comment]
Wasn't able to reproduce and also wasn't able to find scrolling in RHN.  Going to close this.  Reopen/comment if you disagree.

**Comment by mallenexpensify** [comment]
Meant to drop in a vid too

https://github.com/Expensify/App/assets/22508304/2017d96a-227a-41d7-b021-fb30e8be6653
